File: code/cvc5mvn/src/main/java/plaid/eval/Evaluator.java
package plaid.eval;

import plaid.ast.*;

import java.util.*;

public class Evaluator {
    private final Program program;
    public List<Map<Identifier, PreludeExpression>> binding_list;
    
    public Evaluator(Program program){
        this.program = program; 
        binding_list = new LinkedList<>();
        binding_list.add(new HashMap<>());
    }

    /**
     * Evaluate Prelude Expression 
     * @param e Prelude Expression
     * @return Overture 
     */
    public PreludeExpression toOverture(PreludeExpression e) {
        return switch(e){
            // base cases
            case Str w -> new Str(w.str());
            case Num i -> new Num(i.num());
            case Identifier id -> binding_list.getLast().get(id);
            
//            {
//                // if identifier is bound to value 
//                for(Map<Identifier, PreludeExpression> bindings : binding_list){
//                    for(Map.Entry<Identifier, PreludeExpression> entry : bindings.entrySet()){
//                        if(entry.getKey().equals(id)){
//                            // return the value 
//                            yield entry.getValue();
//                        }
//                    }
//                }
//                //yield binding_list.getLast().get(id); // original
//                // otherwise the id itself
//                yield id;
//            } // throws an error if the list doesn't contain the id
           

            case RandomExpr re -> new RandomExpr(toOverture(re.e()));
            case SecretExpr se -> new SecretExpr(toOverture(se.e()));
            case MessageExpr me -> new MessageExpr(toOverture(me.e()));
            case PublicExpr pe -> new PublicExpr(toOverture(pe.e()));
            case OutputExpr oe -> new OutputExpr();

            case TimesExpr te -> new TimesExpr(toOverture(te.e1()), toOverture(te.e2()));
            case PlusExpr pe -> new PlusExpr(toOverture(pe.e1()), toOverture(pe.e2()));
            case ConcatExpr ce -> {
                Str s1 = (Str) toOverture(ce.e1());
                Str s2 = (Str) toOverture(ce.e2());
                yield new Str(s1.str() + s2.str());
            }
            case MinusExpr me -> new MinusExpr(toOverture(me.e()));

            case LetExpr le -> {
                // let y = e1 in e2
                // evalConstraint e1 first
                PreludeExpression e1 = toOverture(le.e1());
                // let y = v in e2
                // substitute v for y in e2

                Map<Identifier, PreludeExpression> bindings = new HashMap<>(binding_list.getLast());
                bindings.put(le.y(), e1);
                binding_list.addLast(bindings);

                PreludeExpression result = toOverture(le.e2());
                binding_list.removeLast();

                yield result;
            }
            case FunctionCallExpr fe -> {
                // evalConstraint the parameters first
                List<PreludeExpression> actualParams = fe
                        .parameters()
                        .stream()
                        .map(this::toOverture)
                        .toList();

                // bind former parameters to actual parameters in function context
                Map<Identifier, PreludeExpression> bindings = new HashMap<>();


                for(int i = 0; i < actualParams.size(); i++){
                    bindings.put(program.resolveExprFunction(fe.fname()).y().get(i), actualParams.get(i));
                }
                binding_list.add(bindings);

                PreludeExpression result = toOverture(program.resolveExprFunction(fe.fname()).e());

                binding_list.removeLast();
                yield result;
            }
            case FieldExpr fe -> {
                TreeMap<Identifier, PreludeExpression> field = new TreeMap<>();
                for(Map.Entry<Identifier, PreludeExpression> element : fe.elements().entrySet()){
                    field.put(element.getKey(), toOverture(element.getValue()));
                }

                yield new FieldExpr(field);
            }
            case FieldSelectExpr fse -> {
                // evalConstraint field expression first
                FieldExpr field = (FieldExpr) toOverture(fse.e());
                yield field.elements().get(fse.l());
            }

            case AtExpr ae -> new AtExpr(toOverture(ae.e1()), toOverture(ae.e2()));
            
            case OTExpr oe -> new OTExpr(toOverture(oe.e1()), toOverture(oe.i1()), toOverture(oe.e2()), toOverture(oe.e3()));
            case OTFourExpr ofe -> new OTFourExpr(toOverture(ofe.s1()), toOverture(ofe.s2()), toOverture(ofe.i1()), toOverture(ofe.e1()), toOverture(ofe.e2()), toOverture(ofe.e3()), toOverture(ofe.e4()));
            default -> throw new IllegalArgumentException("Bad Expression");
        };
    }

    /**
     * Evaluate Prelude Command 
     * @param instr Prelude command
     * @return Overture 
     */
    public PreludeCommand evalInstruction(PreludeCommand instr) {
        return switch (instr) {

            case AssignCommand assignCommand ->
                    new AssignCommand(toOverture(assignCommand.e1()), toOverture(assignCommand.e2()));

            case FunctionCallCommand functionCall -> {
                // evalConstraint actual parameters first
                List<PreludeExpression> actual_parameters =
                        functionCall.parameters().stream().map(this::toOverture).toList();

                // find a function with the same function name
                CommandFunction function = program.resolveCommandFunction(functionCall.fname());

                // map former parameters to actual parameters
                Map<Identifier, PreludeExpression> bindings = new HashMap<>();
                for (int i = 0; i < actual_parameters.size(); i++) {
                    bindings.put(function.typedVariables().get(i).y(), actual_parameters.get(i));
                }

                binding_list.add(bindings);

                // evalConstraint the function body with the actual parameters
                PreludeCommand result = evalInstruction(function.c());
                binding_list.removeLast(); // have to comment out for a constraint to evaluate

                yield result;
            }

            case LetCommand letCommand -> {
                // let y = e in c
                // evalConstraint e
                PreludeExpression v = toOverture(letCommand.e());
                // let y = v in c
                Map<Identifier, PreludeExpression> binding = new HashMap<>(binding_list.getLast());
                binding.put(letCommand.y(), v);
                binding_list.addLast(binding);

                PreludeCommand result = evalInstruction(letCommand.c());
                binding_list.removeLast();
                yield result;
            }

            case CommandList commandList -> {
                List<PreludeCommand> commands = commandList
                        .commands()
                        .stream()
                        .map(this::evalInstruction)
                        .toList();

                yield new CommandList(commands);
            }

            case AssertCommand assertCommand ->
                    new AssertCommand(toOverture(assertCommand.e1()), toOverture(assertCommand.e2()), toOverture(assertCommand.e3()));

            default -> throw new IllegalArgumentException("Bad instruction");

        };
    }

    /**
     * Evalute Constraint Expression 
     * @param e ConstraintExpr
     * @return constraints
     */

    public ConstraintExpr evalConstraint(ConstraintExpr e) {
        return switch (e) {
            case AndConstraintExpr x -> new AndConstraintExpr(evalConstraint(x.e1()), evalConstraint(x.e2()));
            case NotConstraintExpr x -> new NotConstraintExpr(evalConstraint(x.e()));
            case EqualConstraintExpr x -> new EqualConstraintExpr(toOverture(x.e1()), toOverture(x.e2()));
            case FunctionCallExpr x -> {
                ConstraintFunction function = program.resolveConstraintFunction(x.fname());
                List<Identifier> formalParams = function.params();
                List<PreludeExpression> actualParams = x
                        .parameters()
                        .stream()
                        .map(this::toOverture)
                        .toList();

                Map<Identifier, PreludeExpression> bindings = new HashMap<>();
                for (int i = 0; i < actualParams.size(); i++) {
                    bindings.put(formalParams.get(i), actualParams.get(i));
                }

                binding_list.add(bindings);
                ConstraintExpr result = evalConstraint(function.body());
                binding_list.removeLast();

                yield result;
            }
            case TrueConstraintExpr x -> x;
            default -> throw new IllegalArgumentException("Bad constraint");
        };
    }

}

File: code/cvc5mvn/src/main/scala/plaid/logic/Evaluator.scala
package plaid.logic

import plaid.ast.*

import scala.jdk.CollectionConverters.*

/** Support for reducing an AST using variable substitution, expanding function calls, etc. */
case class Evaluator(
  program: Program,
  bindings: Map[Identifier, Expr] = Map()) {

  private def concat(e1: Expr, e2: Expr) = (e1, e2) match
    case (Str(s1), Str(s2)) => Str(s1 + s2)
    case (ConcatExpr(e1, Str(s1)), Str(s2)) => ConcatExpr(e1, Str(s1 + s2))
    case (Str(s1), ConcatExpr(Str(s2), e2)) => ConcatExpr(Str(s1 + s2), e2)
    case _ => ConcatExpr(e1, e2)

  /** Reduces an expression as much as possible in the context of this evaluator's Program and bindings. */
  def expression(e: Expr): Expr = e match
    case str: Str => str
    case num: Num => num
    case id: Identifier => bindings.getOrElse(id, id)
    case RandomExpr(e) => RandomExpr(expression(e))
    case SecretExpr(e) => SecretExpr(expression(e))
    case MessageExpr(e) => MessageExpr(expression(e))
    case PublicExpr(e) => PublicExpr(expression(e))
    case oe: OutputExpr => oe
    case TimesExpr(e1, e2) => TimesExpr(expression(e1), expression(e2))
    case PlusExpr(e1, e2) => PlusExpr(expression(e1), expression(e2))
    case ConcatExpr(e1, e2) => concat(expression(e1), expression(e2))
    case MinusExpr(e)  => MinusExpr(expression(e))
    case LetExpr(y, e1, e2) =>
      val evaluator = copy(bindings = bindings.updated(y, expression(e1)))
      evaluator.expression(e2)
    case FunctionCall(fn, parms) =>
      val f = program.resolveExprFunction(fn)
      val formalParms = f.y
      val actualParms = parms.map(expression)
      val evaluator = copy(bindings = Map.from(formalParms.zip(actualParms)))
      evaluator.expression(f.e)
    case FieldExpr(elements) =>
      val field = scala.collection.immutable.TreeMap(elements.asScala.toSeq.map((k,v) => k -> expression(v)):_*)
      FieldExpr(java.util.TreeMap(field.asJava))

    case FieldSelectExpr(e, l) =>
      val field = expression(e).asInstanceOf[FieldExpr]
      field.elements.get(l)

    case AtExpr(e1, e2) => AtExpr(expression(e1), expression(e2))

    case OTExpr(e1, i1, e2, e3) =>
      OTExpr(expression(e1), expression(i1), expression(e2), expression(e3))

    case OTFourExpr(s1, s2, i1, e1, e2, e3, e4) =>
      OTFourExpr(
        expression(s1), expression(s2), expression(i1),
        expression(e1), expression(e2), expression(e3), expression(e4))

  /** Reduces a command as much as possible in the context of this evaluator's Program and bindings. */
  def command(cmd: Cmd): Cmd = cmd match
    case AssignCmd(e1, e2) => AssignCmd(expression(e1), expression(e2))
    case CallCmd(fn, parms) =>
      val f = program.resolveCommandFunction(fn)
      val formalParms = f.typedVariables.map(_.y)
      val actualParms = parms.map(expression)
      val evaluator = copy(bindings = Map.from(formalParms.zip(actualParms)))
      evaluator.command(f.c)
    case LetCmd(y, e, c) =>
      val evaluator = copy(bindings = bindings.updated(y, expression(e)))
      evaluator.command(c)
    case ListCmd(c1, c2) => ListCmd(command(c1), command(c2))
    case AssertCmd(e1, e2, e3) => AssertCmd(expression(e1), expression(e2), expression(e3))
    case _ => throw new IllegalArgumentException("Bad instruction")

  /** Reduces a constraint as much as possible in the context of this evaluator's Program and bindings. */
  def constraint(c: Constraint): Constraint = c match
    case AndConstraint(e1, e2) => AndConstraint(constraint(e1), constraint(e2))
    case NotConstraint(e) => NotConstraint(constraint(e))
    case EqualConstraint(e1, e2) => EqualConstraint(expression(e1), expression(e2))
    case FunctionCall(fn, parms) =>
      val f = program.resolveConstraintFunction(fn)
      val formalParams = f.params
      val actualParams = parms.map(expression)
      val evaluator = copy(bindings = Map.from(formalParams.zip(actualParams)))
      evaluator.constraint(f.body)
    case x: TrueConstraint => x
    case _ => throw new IllegalArgumentException("Bad constraint")
}
